libsolidity/analysis/TypeChecker.cpp

```diff
@@ -4035,24 +4065,73 @@ void TypeChecker::endVisit(Literal const& _literal)
 						),
 						"Functions that take 2 arguments can only be used as literal suffixes for rational numbers."
 					);
-				else
+				else if (
+					dynamic_cast<IntegerType const*>(suffixFunctionType->parameterTypes().at(0)) &&
+					dynamic_cast<IntegerType const*>(suffixFunctionType->parameterTypes().at(1))
+				)
 				{
 					auto&& [mantissa, exponent] = literalRationalType->mantissaExponent();
 					// This was already validated in visit(Literal) but the error is not fatal.
 					if (!mantissa || !exponent)
 						solAssert(!m_errorReporter.errors().empty());
-					else if (
-						!mantissa->isImplicitlyConvertibleTo(*suffixFunctionType->parameterTypes().at(0)) ||
-						!exponent->isImplicitlyConvertibleTo(*suffixFunctionType->parameterTypes().at(1))
-					)
-						// TODO: Is this triggered when the argument is out of range? Test.
-						parameterTypeMessage = "The type of the literal cannot be converted to the parameters of the suffix function.";
+					// TODO: Is this triggered when the argument is out of range? Test.
+					else
+					{
+						vector<string> mantissaOrExponentErrorMessages;
+						if (!mantissa->isImplicitlyConvertibleTo(*suffixFunctionType->parameterTypes().at(0)))
+							mantissaOrExponentErrorMessages.emplace_back(
+								fmt::format(
+									"The mantissa is out of range of type {}.",
+									suffixFunctionType->parameterTypes().at(0)->toString(true)
+								)
+							);
+						if (!exponent->isImplicitlyConvertibleTo(*suffixFunctionType->parameterTypes().at(1)))
+							mantissaOrExponentErrorMessages.emplace_back(
+								fmt::format(
+									"The exponent is out of range of type {}.",
+									suffixFunctionType->parameterTypes().at(1)->toString(true)
+								)
+							);
+
+						if (!mantissaOrExponentErrorMessages.empty())
+							parameterTypeMessage =
+								"This number cannot be decomposed into a mantissa and decimal exponent "
+								"that fit the range of parameters of the suffix function. " +
+								joinHumanReadable(mantissaOrExponentErrorMessages, " ");
+					}
 				}
+				else
+					// visit(FunctionDefinition) should have spotted if any of the parameters is not IntegerType
+					solAssert(!m_errorReporter.errors().empty());
 			}
 			else if (suffixFunctionType->parameterTypes().size() == 1)
 			{
 				if (!literalType->isImplicitlyConvertibleTo(*suffixFunctionType->parameterTypes().at(0)))
-					parameterTypeMessage = "The type of the literal cannot be converted to the parameter of the suffix function.";
+				{
+					string conversionErrorMessage;
+					if (
+						dynamic_cast<IntegerType const*>(suffixFunctionType->parameterTypes().at(0)) &&
+						literalRationalType
+					)
+						conversionErrorMessage = "The number is out of range of";
+					else
+					{
+						if (literalRationalType)
+							conversionErrorMessage = "The number";
+						else if (dynamic_cast<AddressType const*>(literalType))
+							conversionErrorMessage = "The address";
```

Comment:
This is good, but when reading the code here it may not be clear why we're singling out addresses without seeing the result in the test case. I'd add a comment to explain it:
```suggestion
						else if (dynamic_cast<AddressType const*>(literalType))
							// Address literals and hex numbers are not easy to distinguish without counting the digits.
							// The user may not even realize we see it as an address. Let's point that out in the message for clarity.
							conversionErrorMessage = "The address";
```